Enable devirtualization in more scenarios on crossgen2

davidwrighton opened the pull request:
Address deficiencies in current devirtualization infrastructure

- Remove the responsibility of creating a CORINFO_RESOLVED_TOKEN structure from the JIT and make it a responsibility of the VM side of the jit interface.
  - This enables the component (crossgen2) which has deeper understanding of the requirements here to correctly handle scenarios that would otherwise require expressing crossgen2 specific details across the jit interface.
- Add a new set of fixups (`READYTORUN_FIXUP_Check_VirtualFunctionOverride` and `READYTORUN_FIXUP_Verify_VirtualFunctionOverride`) these are used to validate that the behavior of the runtime and crossgen2 compiler is equivalent for a virtual resolution event
  - `READYTORUN_FIXUP_Check_VirtualFunctionOverride` will ensure that the virtual resolution decision is the same at crossgen2 time and runtime, and if the decision differs, any generated code affected by the decision will not be used.
  - `READYTORUN_FIXUP_Verify_VirtualFunctionOverride` will perform the same checks as `READYTORUN_FIXUP_Check_VirtualFunctionOverride`, but if it fails the check, the process will be terminated with a fail-fast. It is intended for use under the `--verify-type-and-field-layout` stress mode.
  - Currently only the `READYTORUN_FIXUP_Verify_VirtualFunctionOverride` is actually generated, and it is only generated when using the `--verify-type-and-field-layout` switch to crossgen2. Future work will identify if there are scenarios where we need to generate the `READYTORUN_FIXUP_Check_VirtualFunctionOverride` flag. One area of possible concern is around covariant returns, another is around handling of type equivalence.
- In order to express the fixup signature for the VirtualFunctionOverride fixups, a new flag has been added to `ReadyToRunMethodSigFlags`. `READYTORUN_METHOD_SIG_UpdateContext` will allow the method signature to internally specify the assembly which is associated with the method token, instead of relying on the ambient context.
- R2RDump and the ReadyToRun format documentation have been updated with the details of the new fixups/flags.
- Update the rules for handling unboxing stubs
  - See #51918 for details. This adds a new test, as well as proper handling for unboxing stubs to match the JIT behavior
  - Also revert #52605, which avoided the problem by simply disabling devirtualization in the presence of structs
- Adjust the rules for when it is legal to devirtualize and maintain version resiliency
  - The VersionsWithCode and VersionsWithType rules are unnecessarily restrictive.
  - Instead Validate that the metadata is safely checkable, and rely on the canInline logic to ensure that no IL that can't be handled is inlined.
  - This also involved adding a check that the chain of types from the implementation type to the declaration method table type is within the version bubble.
  - And changing the `VersionsWithType` check on the implementation type, to a new `VersionsWithTypeReference` check which can be used to validate that the type can be referred to, in combination with using `VersionsWithType` on the type definition.
  - By adjusting the way that the declMethod is referred to, it becomes possible to use the declMethod without checking the full method is `VersionsWithCode`, and it just needs a relationship to version matching code.
- In `resolveVirtualMethod` generate the `CORINFO_RESOLVED_TOKEN` structures for the jit
  - In particular we are now able to resolve to methods where the decl method is the resolution result but is not within the version bubble itself. This can happen if we can prove that the decl method is the only method which can possibly implement a virtual.
- Add support for devirtualization reasons to crossgen2
- Port all devirtualization abort conditions to crossgen2 from runtime that were not already present
- Fix devirtualization from a canonical virtual method when the actual implementation is more exact
- Fix variant interface override scenario where there is an interface that requires implementation of the variant interface as well as the variant interface itself.

MichalStrehovsky committed:
Don't forget to pass instantiation context after devirtualizing

This is a difference between CoreCLR and the managed type system - the result of interface method resolution is going to be a canonical method on the CoreCLR side, but exact on the managed side.

So this implementation is slightly different from how it looks like on the VM side. We slightly touched on the difference in https://github.com/dotnet/runtime/pull/45744 but there was more to it.

davidwrighton committed:
Re-enable devirtualization in crossgen2 for structs, and increase the scenarios in which it is used

davidwrighton committed:
Enable a new pair of check/verify fixups for validating virtual function override behavior
- Enable use of these fixups for devirtualization scenarios
- when --verify-type-and-field-layout is specified, the verification variant of the fixup is used
- Update documentation and r2rdump utility

davidwrighton committed:
Fix most obvious issues

davidwrighton committed:
Handle canonically compatible cases

davidwrighton committed:
Add more logic to our devirtualization logic to be ever more certain that we won't devirtualize outside of the version bubble

davidwrighton committed:
Tweak logic to record tokens for later use. I'm not convinced this is ok though, as I think type forwarders can make this R2R unsafe

davidwrighton committed:
Rework api surface of devirtualizer to avoid the need to cache token info

davidwrighton committed:
Don't generate the UpdateContext record for signatures used in instrumentation and instantiated tables as the owner type is sufficient for now.
- Also, set resolvedTokenDevirtualizedUnboxedMethod correctly

davidwrighton committed:
Fix R2RDump decoding in composite images
Use correct token

davidwrighton committed:
format patch

davidwrighton committed:
Decode exact method for decl method

davidwrighton committed:
Tweak slot unification to properly ignore name+sig matches on the same type as a MethodImpl override

davidwrighton committed:
Fix issue where we could devirtualize to a method that could not be referred to

davidwrighton committed:
Validate that the dec method can be represented in a methodsig before it is used in such a way

davidwrighton committed:
Update version resiliency rules to be the more precise ones that have been determined via analysis

davidwrighton committed:
Now that we have a slightly more capable VersionWith predicate for type signature generation, use it in the InstrumentationDataTableNode

davidwrighton committed:
- Allow locally defined structs to devirtualize on interface calls made on non-locally defined interfaces.
- decl vs impl check done on exact result, not on canonicalized result

davidwrighton committed:
- Ensure all of the devirtualization abort conditions from the runtime are implemented in crossgen2
- Provide devirtualization details for all possible ways devirt can fail in crossgen2 (This is more detailed than the crossgen output)

davidwrighton commented:
@dotnet/crossgen-contrib

AndyAyersMS commented on src/coreclr/jit/fginline.cpp:
```diff
@@ -731,7 +731,7 @@ Compiler::fgWalkResult Compiler::fgLateDevirtualization(GenTree** pTree, fgWalkD
             const bool             isLateDevirtualization = true;
             bool explicitTailCall = (call->AsCall()->gtCallMoreFlags & GTF_CALL_M_EXPLICIT_TAILCALL) != 0;
-            comp->impDevirtualizeCall(call, &method, &methodFlags, &context, nullptr, isLateDevirtualization,
+            comp->impDevirtualizeCall(call, NULL, &method, &methodFlags, &context, nullptr, isLateDevirtualization,
```
```suggestion
            comp->impDevirtualizeCall(call, nullptr, &method, &methodFlags, &context, nullptr, isLateDevirtualization,
```

AndyAyersMS commented on src/coreclr/jit/importer.cpp:
```diff
@@ -20683,6 +20684,7 @@ bool Compiler::IsMathIntrinsic(GenTree* tree)
 //     is not exactly known, and there is a plausible guess for the type.
 //
 void Compiler::impDevirtualizeCall(GenTreeCall*            call,
+                                   CORINFO_RESOLVED_TOKEN* pResolvedToken,
```
Can you add a comment for the new arg to the block just above?

AndyAyersMS commented on src/coreclr/jit/importer.cpp:
```diff
@@ -21666,9 +21656,11 @@ void Compiler::considerGuardedDevirtualization(
+    dvInfo.objClass                    = likelyClass;
+    dvInfo.context                     = *pContextHandle;
+    dvInfo.exactContext                = *pContextHandle;
+    dvInfo.pResolvedTokenVirtualMethod = NULL;
```
```suggestion
    dvInfo.pResolvedTokenVirtualMethod = nullptr;
```

AndyAyersMS commented on src/coreclr/tools/aot/ILCompiler.ReadyToRun/Compiler/DependencyAnalysis/ReadyToRun/DevirtualizationManager.cs:
```diff
@@ -24,44 +25,197 @@ public override bool IsEffectivelySealed(MethodDesc method)
+            //    jit interface logic after the logic that executes here.
+            //
+            // ImplType checking
+            // 1. At all times the metadata definition of the implmenetation type must version with the application.
```
```suggestion
            // 1. At all times the metadata definition of the implementation type must version with the application.
```

AndyAyersMS reviewed (commented):
Looked over the jit changes, which generally look good, and skimmed the rest, much of which I'm not going to be able to review effectively.

This will need a new jit guid.

There will need to be some updates to superpmi's handling of `resolveVirtualMethod`; you can likely use the existing helpers for resolved tokens (eg `SpmiRecordsHelper::StoreAgnostic_CORINFO_RESOLVED_TOKENout`) Seems like you might need a novel combo of the in and out helpers though, as in this case all the fields of the resolved token are outputs...?

trylek commented on src/tests/JIT/opt/Devirtualization/GitHub_51918.cs:
```diff
@@ -0,0 +1,58 @@
+    public int _v1;
+    public int _v2;
+
+    public GenericClass()
```
I can find the original issue but it would be extremely helpful if you could add at least a three-line comment describing the way the test is supposed to work and how it detects bad behavior - in my recent work on fixing the various regression tests with Crossgen2 I've always had to spend quite a bit of time poring over the particular test to understand what it's supposed to be doing especially when it's exercising some subtle behavioral differences.

trylek reviewed (approved):
Looks beautiful to me, thank you!

davidwrighton committed:
- Use variant interface resolution in devirtualization algorithm
- Fix variant interface dispatch bug where the variant interface is implemented by another interface associated with the type.
- Handle default interface methods
  - Except for variant dispatch, detect those, and fail the devirtualization

davidwrighton commented on src/coreclr/vm/zapsig.cpp:
```diff
@@ -801,6 +810,22 @@ MethodDesc *ZapSig::DecodeMethod(Module *pInfoModule,
+        uint32_t updatedModuleIndex;
+        IfFailThrow(sig.GetData(&updatedModuleIndex));
+#ifdef FEATURE_MULTICOREJIT
+        if (pZapSigContext->externalTokens == ZapSig::MulticoreJitTokens)
```
MulticoreJitTokens is used for the MulticoreJit feature, which is an odd profile guided optimization feature that works on a local machine. We don't need to worry about it in crossgen2.

davidwrighton commented on src/coreclr/tools/aot/ILCompiler.ReadyToRun/Compiler/ReadyToRunCompilationModuleGroupBase.cs:
```diff
@@ -399,6 +416,93 @@ private bool ComputeTypeVersionsWithCode(TypeDesc type)
 
+        private bool ComputeTypeReferenceVersionsWithCode(TypeDesc type)
+        {
+            switch(type.Category)
```
Its not quite the same as ComputeInstantiationVersionsWithCode. This one includes Void, but I've used the IsPrimitive and IsVoid predicates, and added a comment.

davidwrighton committed:
Code review and formatting

davidwrighton commented on src/coreclr/tools/aot/ILCompiler.ReadyToRun/JitInterface/CorInfoImpl.ReadyToRun.cs:
```diff
@@ -247,6 +247,24 @@ public void AppendMangledName(NameMangler nameMangler, Utf8StringBuilder sb)
 
+        public override string ToString()
+        {
+            StringBuilder debuggingName = new StringBuilder();
```
Its not quite the same, in particular, ToString() doesn't have access to a NameMangler.

davidwrighton committed:
Update jit guid, and superpmi

MichalStrehovsky commented on src/coreclr/tools/Common/JitInterface/CorInfoImpl.cs:
```diff
@@ -1160,23 +1161,132 @@ private bool resolveVirtualMethod(CORINFO_DEVIRTUALIZATION_INFO* info)
+                result.tokenContext = jitInterface.contextFromMethod(method);
+                result.token = methodWithToken.Token.Token;
+                if (methodWithToken.Token.TokenType != CorTokenType.mdtMethodDef)
+                    throw new NotSupportedException();
```
This will tear down the compiler process. Do we want that to happen? Consider RequiresRuntimeJitException (not fatal) or CodeGenerationFailedException (that one can be fatal if we're not in the resilient mode).

MichalStrehovsky commented on src/coreclr/tools/Common/TypeSystem/IL/EcmaMethodIL.cs:
```diff
@@ -11,7 +11,10 @@
 {
-    public sealed partial class EcmaMethodIL : MethodIL
+    // Marker interface implemented by EcmaMethodIL and EcmaMethodILScope
+    public interface IEcmaMethodIL { }
```
Nit: might be faster to just do the two `is` checks instead of testing for the interface. Both EcmaMethodIL and EcmaMethodILScope are sealed so an `is` is quick.

davidwrighton commented on src/coreclr/tools/Common/Compiler/DevirtualizationManager.cs:
```diff
@@ -84,20 +85,91 @@ protected virtual MethodDesc ResolveVirtualMethod(MethodDesc declMethod, DefType
+                                {
+                                    // If we devirtualized into a default interface method on a generic type, we should actually return an
+                                    // instantiating stub but this is not happening.
+                                    // Making this work is tracked by https://github.com/dotnet/runtime/issues/9588
```
Well, its the same problem as devirtualizing to a struct method, which is something that @AndyAyersMS has actually made work fairly recently. I suspect that the comment is somewhat out of date, but I felt it was useful to have a matching comments in both crossgen2 and coreclr vm.

davidwrighton committed:
Attempt to diagnose outerloop issues by adding instrumentation and running different tasks during outerloop

davidwrighton committed:
Fixup issues with virtual resolution of normal vtable methods and a canonicallly compatible vtable devirtualization, and go back to normal yml for the crossgen2 outerloop

davidwrighton commented on src/coreclr/tools/Common/Compiler/DevirtualizationManager.cs:
```diff
@@ -53,16 +54,16 @@ public virtual bool IsEffectivelySealed(MethodDesc method)
         /// might have to be treated as an unboxing thunk by the caller.
         /// </remarks>
-        public MethodDesc ResolveVirtualMethod(MethodDesc declMethod, TypeDesc implType)
+        public MethodDesc ResolveVirtualMethod(MethodDesc declMethod, TypeDesc implType, ref CORINFO_DEVIRTUALIZATION_DETAIL devirtualizationDetail)
```
Yes. Turns out not doing this was masking a bug in the devirtualilzation algorithm

davidwrighton committed:
Address code review feedback
